trackpy/linking.py

```diff
@@ -744,139 +768,231 @@ def link_iter(levels, search_range, memory=0,
+                targeted_predictor = functools.partial(self.predictor, t_next)
+                prev_hash.rebuild(coord_map=targeted_predictor) # Rewrite positions
+
+            # Now we can process the new particles.
+            # Make a Hash / Tree for the destination level.
+            if self.neighbor_strategy == 'BTree':
+                cur_hash = self.hash_generator()
+                for p in cur_set:
+                    cur_hash.add_point(p)
+            elif self.neighbor_strategy == 'KDTree':
+                cur_hash = TreeFinder(cur_level)
+
+            # Set up attributes for keeping track of possible connections.
             for p in cur_set:
-                cur_hash.add_point(p)
-        elif neighbor_strategy == 'KDTree':
-            cur_hash = TreeFinder(cur_level)
+                p.back_cands = []
+                p.forward_cands = []
 
-        # Set up attributes for keeping track of possible connections.
-        for p in cur_set:
-            p.back_cands = []
-            p.forward_cands = []
+            # Sort out what can go to what.
+            assign_candidates(cur_level, prev_hash, self.search_range,
+                              self.neighbor_strategy)
 
-        # Sort out what can go to what.
-        assign_candidates(cur_level, prev_hash, search_range,
-                          neighbor_strategy)
+            # sort the candidate lists by distance
+            for p in cur_set:
+                p.back_cands.sort(key=lambda x: x[1])
+            for p in prev_set:
+                p.forward_cands.sort(key=lambda x: x[1])
 
-        # sort the candidate lists by distance
-        for p in cur_set:
-            p.back_cands.sort(key=lambda x: x[1])
-        for p in prev_set:
-            p.forward_cands.sort(key=lambda x: x[1])
+            # Note that this modifies cur_set, prev_set, but that's OK.
+            spl, dpl = self.assign_links(cur_set, prev_set, self.search_range)
 
-        new_mem_set = set()
-        # while there are particles left to link, linkge the repo
-        while len(cur_set) > 0:
-            p = cur_set.pop()
+            new_mem_set = set()
+            for sp, dp in zip(spl, dpl):
+                # Do linking
+                if sp is not None and dp is not None:
+                    sp.track.add_point(dp)
+                    _maybe_remove(self.mem_set, sp)
+                elif sp is None:
+                    # if unclaimed destination particle, a track is born!
+                    self.track_lst.append(self.track_cls(dp))
+                elif dp is None:
+                    # add the unmatched source particles to the new
+                    # memory set
+                    new_mem_set.add(sp)
+
+                # Clean up
+                if dp is not None:
+                    del dp.back_cands
+                if sp is not None:
+                    try:
+                        del sp.forward_cands
```

Comment:
I remember adding this for a reason, but it may have been temporary until I could fix a bug elsewhere. I can't see a reason for it now, and when I comment out the "try" the tests still pass. So I'm going to take it out and we'll hope for the best — the alternative is having it there forever :). Good catch!